controller/volume_controller.go

```diff
@@ -1553,6 +1655,260 @@ func getRebuildingReplicaCount(e *longhorn.Engine) int {
+		}
+	}()
+
+	setting, err := vc.getAutoBalancedReplicasSetting(v)
+	enabled := []string{
+		string(types.ReplicaAutoBalanceLeastEffort),
+		string(types.ReplicaAutoBalanceBestEffort),
+	}
+	if !util.Contains(enabled, string(setting)) || err != nil {
+		return 0
+	}
+
+	if v.Status.Robustness != types.VolumeRobustnessHealthy {
+		log.Debugf("Cannot auto-balance volume in %s state", v.Status.Robustness)
+		return 0
+	}
+
+	var adjustCount int
+	adjustCount, _, err = fnCount(v, rs)
+	if err != nil {
+		return 0
+	}
+	log.Debugf("Found %v replica candidate for auto-balance", adjustCount)
+	return adjustCount
+}
+
+func (vc *VolumeController) replicaAutoBalanceBestEffort(v *longhorn.Volume,
+	rs map[string]*longhorn.Replica,
+	fnCount replicaAutoBalanceCount) (int, []string) {
+	log := getLoggerForVolume(vc.logger, v).WithField("replicaAutoBalanceOption", types.ReplicaAutoBalanceBestEffort)
+
+	var err error
+	defer func() {
+		if err != nil {
+			log.WithError(err).Warn("Skip replica auto-balance")
+		}
+	}()
+
+	setting, err := vc.getAutoBalancedReplicasSetting(v)
+	if setting != types.ReplicaAutoBalanceBestEffort || err != nil {
+		return 0, nil
+	}
+
+	if v.Status.Robustness != types.VolumeRobustnessHealthy {
+		log.Debugf("Cannot auto-balance volume in %s state", v.Status.Robustness)
+		return 0, nil
+	}
+
+	var unusedCount int
+	unusedCount, idOverlays, err := fnCount(v, rs)
+	if unusedCount == 0 && len(idOverlays) <= 1 {
+		return 0, nil
+	}
+
+	var topCountIDRNames []string
+	var rOverlapTopCount int
+	var rOverlapLowCount int
+	for _, rNames := range idOverlays {
+		rNameCount := len(rNames)
+		if len(rNames) > len(topCountIDRNames) {
+			topCountIDRNames = rNames
+			rOverlapTopCount = rNameCount
+		}
+		if rOverlapLowCount == 0 || rNameCount < rOverlapLowCount {
+			rOverlapLowCount = rNameCount
+		}
+	}
+	adjustCount := rOverlapTopCount - rOverlapLowCount
+	if adjustCount < 2 {
+		return 0, nil
+	}
+
+	log.Debugf("Found %v replicas to balance: %v", adjustCount, topCountIDRNames)
+	return adjustCount, topCountIDRNames
+}
+
+func (vc *VolumeController) replicaAutoBalanceZoneCount(v *longhorn.Volume, rs map[string]*longhorn.Replica) (int, map[string][]string, error) {
+	log := getLoggerForVolume(vc.logger, v).WithField("replicaAutoBalanceType", "zone")
+
+	readyNodes, err := vc.ds.ListReadyAndSchedulableNodes()
+	if err != nil {
+		return 0, nil, err
+	}
+
+	var usedZones []string
+	var usedNodes []string
+	zoneDuplicates := make(map[string][]string)
+	for _, r := range rs {
+		if r.Status.CurrentState != types.InstanceStateRunning {
+			continue
+		}
+
+		node, exist := readyNodes[r.Spec.NodeID]
+		if !exist {
+			log.Debugf("Replica %v on node %v not count for auto-balance, could get evicted", r.Spec.NodeID, r.Name)
+			continue
+		}
+
+		nZone := node.Status.Zone
+		_, exist = zoneDuplicates[nZone]
+		if exist {
+			zoneDuplicates[nZone] = append(zoneDuplicates[nZone], r.Name)
+		} else {
+			zoneDuplicates[nZone] = nil
```

Comment:
Do you mean `zoneDuplicates[nZone] = []string{}`? I think the list records extra replicas only.

Similar below

Reply:
Addressed. and yes the list records only extra replicas so we preserve at least one.